[CLANG] Add warning when INF or NAN are used in a binary operation or as function argument in fast math mode.

zahiraam opened the pull request:
Check for operations using INF or NaN when in ffast-math mode and generate a warning.

zahiraam committed:
[CLANG] Add warning when comparing to INF or NAN in fast math mode.

andykaylor commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -6771,6 +6771,9 @@ def warn_pointer_sub_null_ptr : Warning<
   "comparing floating point with == or != is unsafe">,
   InGroup<DiagGroup<"float-equal">>, DefaultIgnore;
+def warn_fast_floatingpoint_eq : Warning<
+  "explicit comparison with %0 in fast floating point mode">,
```
Maybe this message should either be more specific or not mention the floating point mode at all. Since this will trigger with just "-fno-honor-nans" or "-fno-honor-infinities" the message is somewhat off target. I guess either of these is a fast floating point mode, but the message makes me think of full fast-math mode, which isn't necessarily the case.

```suggestion
  "explicit comparison with %0 when the program is assumed to not use or produce %0">,
```

andykaylor commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2245,6 +2246,11 @@ Sema::CheckBuiltinFunctionCall(FunctionDecl *FDecl, unsigned BuiltinID,
   case Builtin::BI__builtin_islessequal:
   case Builtin::BI__builtin_islessgreater:
   case Builtin::BI__builtin_isunordered:
+    if (BuiltinID == Builtin::BI__builtin_isunordered) {
```
Should this check be in SemaBuiltinUnorderedCompare()?

andykaylor commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2267,6 +2273,16 @@ Sema::CheckBuiltinFunctionCall(FunctionDecl *FDecl, unsigned BuiltinID,
   case Builtin::BI__builtin_signbit:
   case Builtin::BI__builtin_signbitf:
   case Builtin::BI__builtin_signbitl:
+    FPO = TheCall->getFPFeaturesInEffect(getLangOpts());
```
Should these checks be in SemaBuiltinFPClassification()?

zahiraam replied:
Done.

andykaylor commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -13846,6 +13880,37 @@ Sema::CheckReturnValExpr(Expr *RetValExp, QualType lhsType,
+/// Diagnose comparison to NAN or INFINITY in fast math modes.
+/// The comparison to NaN or INFINITY is always false in
+/// fast modes: float evaluation will not result in inf or nan.
+void Sema::CheckInfNaNFloatComparison(SourceLocation Loc, Expr *LHS, Expr *RHS,
```
This line is 81 characters long.

zahiraam replied:
It's 79 chars long.

andykaylor replied:
Oh, sorry. I must have copied the "+ " when I checked it. It just looked longer than the surrounding text.

andykaylor commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -13846,6 +13880,37 @@ Sema::CheckReturnValExpr(Expr *RetValExp, QualType lhsType,
+  bool NoHonorInfs = FPO.getNoHonorInfs();
+  llvm::APFloat Value(0.0);
+  bool IsConstant;
+  IsConstant = !LHS->isValueDependent() &&
```
You could use a lambda here to avoid code duplication for LHS and RHS checking.

zahiraam replied:
Done.

andykaylor commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -13044,9 +13044,12 @@ static QualType checkArithmeticOrEnumeralCompare(Sema &S, ExprResult &LHS,
-  if (Type->hasFloatingRepresentation())
+  if (Type->hasFloatingRepresentation()) {
+    // Check for comparisons to NAN or INFINITY in fast math mode.
+    S.CheckInfNaNFloatComparison(Loc, LHS.get(), RHS.get(), Opc);
```
Should this be done in CheckFloatComparison()?

zahiraam replied:
Done.

andykaylor reviewed (commented):
Expanding the scope a bit, it would also be useful to have warnings for constant NaN or Inf values passed as arguments or used in binary operations.

zahiraam committed:
Addressed review comments.

zahiraam commented:
> Expanding the scope a bit, it would also be useful to have warnings for constant NaN or Inf values passed as arguments or used in binary operations.

Added that.

arsenm commented on clang/test/Sema/warn-fp-fast-compare.cpp:
```diff
@@ -0,0 +1,245 @@
+}
+#define NAN (__builtin_nanf(""))
+#define INFINITY (__builtin_inff())
+
```
What about through std::numeric_limits<T>::infinity?

zahiraam replied:
Done.

AaronBallman commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -6771,6 +6771,9 @@ def warn_pointer_sub_null_ptr : Warning<
   "comparing floating point with == or != is unsafe">,
   InGroup<DiagGroup<"float-equal">>, DefaultIgnore;
+def warn_fast_floatingpoint_eq : Warning<
+  "explicit comparison with %0 when the program is assumed to not use or produce %0">,
```
```suggestion
  "explicit comparison with %select{infinity|NaN}0 will always return '%select{false|true}1' because %select{infinity|NaN}0 will not be produced according to the currently enabled floating-point options">,
```
How about something along these lines? I mostly want to replace the `%0` with a `select` so we don't have to pass in constant strings at the use sites, and I want to explicitly tell the user whether the comparison will always be true or always be false.

zahiraam replied:
Not sure this makes sense. Since I am also adding this warning to things like a * INF/NaN or foo(INF).

zahiraam replied:
If it's ok I will use "use of infinity|NaN will always" only because the warning is not triggered only for comparison but for binary operators using inf/nan and function calls with inf/nan arguments.

zahiraam replied:
@AaronBallman WDYT?

AaronBallman commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -9090,10 +9092,16 @@ bool Sema::SemaBuiltinVAStartARMMicrosoft(CallExpr *Call) {
   if (checkArgCount(*this, TheCall, 2))
     return true;
 
+  if (BuiltinID == Builtin::BI__builtin_isunordered) {
+    if (TheCall->getFPFeaturesInEffect(getLangOpts()).getNoHonorNaNs())
+      Diag(TheCall->getBeginLoc(), diag::warn_fast_floatingpoint_eq)
+          << "NaN" << TheCall->getSourceRange();
+  }
```
```suggestion
  if (BuiltinID == Builtin::BI__builtin_isunordered &&
      TheCall->getFPFeaturesInEffect(getLangOpts()).getNoHonorNaNs())
        Diag(TheCall->getBeginLoc(), diag::warn_fast_floatingpoint_eq)
            << "NaN" << TheCall->getSourceRange();
```

AaronBallman reviewed (commented):
The changes also need a release note.

zahiraam committed:
Addressed review comments.

andykaylor commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -6771,6 +6771,11 @@ def warn_pointer_sub_null_ptr : Warning<
   "comparing floating point with == or != is unsafe">,
   InGroup<DiagGroup<"float-equal">>, DefaultIgnore;
+def warn_fast_floatingpoint_eq : Warning<
+  "use of %select{infinity|NaN}0 will always be 'false' because "
```
This doesn't make sense to me. What does it mean for a use to be false?

Maybe we're trying to wrap too much into a single diagnostic. For comparisons, the result of the comparison will be constant, so it makes sense to have a warning that tells the user what that constant result will be. However, when a constant NaN or infinity is passed as an argument to a function call, the situation is a bit more complicated. In that case, the optimizer can treat the input as poison. What effect that will have depends on the content of the called function. It might do nothing. It might cause the entire function call to be erased.

zahiraam replied:
Yea! not happy with this either. 
May be saying something like this to warn the user the existence of inf/nan use: "use of inf/nan when the program will not be producing inf/nan according to the currently enabled floating-point options"?

zahiraam committed:
Fixed the diagnostic message.

zahiraam committed:
Fix format.

zahiraam committed:
Merge remote-tracking branch 'origin/main' into WarnInfNanFastMode

AaronBallman commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -6784,6 +6784,12 @@ def warn_pointer_sub_null_ptr : Warning<
   "comparing floating point with == or != is unsafe">,
   InGroup<DiagGroup<"float-equal">>, DefaultIgnore;
+def warn_fast_floatingpoint_eq : Warning<
+  "%select{explicit comparison with|use of}0 %select{infinity|nan}1 "
+  "%select{will always return 'false'|as a function argument will not always be interpreted as such}0 "
+  "because %select{infinity|nan}1 will not be produced according to the "
+  "currently enabled floating-point options">,
```
Ah, now that I better understand the scope of the functionality, I think we should go with:

`use of %select{infinity|NaN}0 results in undefined behavior due to the currently enabled floating-point options`

and move the diagnostic out of the tautological compare warning group. You should also update ExprConstant.cpp to reject use of inf/nan in a constant expression with the options disabled.

AaronBallman commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -13861,6 +13897,36 @@ Sema::CheckReturnValExpr(Expr *RetValExp, QualType lhsType,
+/// Diagnose comparison to NAN or INFINITY in fast math modes.
+/// The comparison to NaN or INFINITY is always false in
+/// fast modes: float evaluation will not result in inf or nan.
+void Sema::CheckInfNaNFloatComparison(SourceLocation Loc, Expr *LHS, Expr *RHS,
```
The basic idea I have is that it's UB to generate a constant infinity or NaN if infinity/NaN is disabled. Instead of performing special logic for float comparisons, I think the logic should live in `Sema::ActOnNumericConstant()` so that we capture any time a float or infinity constant is used. e.g. `inf + 1` should also diagnose.

Then the tautological comparison component to this could be implemented by looking for calls to `__builtin_isinf` or `__builtin_isnan` and claim those are tautologically false (as a follow-up patch). WDYT?

zahiraam replied:
I have actually tried that but I noticed that `getSpellingOfSingleCharacterNumericConstant` of inf/nan will both return "0" so I will have no way to distinguish between inf + 1 and 0 + 1 and both will generate a warning. Unless there is other way of finding out that the token is nan/inf?

AaronBallman replied:
Why are we on that path in the first place? The token length shouldn't be one character long for either `inf` or `nan` right? I was thinking you should be able to call `NumericLiteralParser::GetFloatValue()` within the `} else if (Literal.isFloatingLiteral()) {` block to get an `APFloat` that you can call `isInfinity()` or `isNaN()` on and then diagnose at that point.

zahiraam replied:
@AaronBallman Thanks.

zahiraam committed:
After discussion with Aaron separated the warning so one is triggered
when INFINITY/NAN macros are used and the other when __builtin_inf()
and __builtin_nan() are used.

zahiraam committed:
Fixed the diagnostics and RN.

AaronBallman commented on clang/include/clang/Basic/DiagnosticCommonKinds.td:
```diff
@@ -70,6 +70,10 @@ def warn_pragma_debug_missing_argument : Warning<
   "unexpected argument to debug command">, InGroup<IgnoredPragmas>;
 
+def warn_fp_nan_inf_when_disabled : Warning<
+  "use of %select{infinity|NaN}0 %select{|via a macro}1 results in an undefined behavior "
```
```suggestion
  "use of %select{infinity|NaN}0%select{| via a macro}1 results in undefined behavior "
```
How about this?

AaronBallman commented on clang/include/clang/Lex/Preprocessor.h:
```diff
@@ -2835,6 +2835,9 @@ class Preprocessor {
         !SourceMgr.isInMainFile(Identifier.getLocation()))
       emitRestrictExpansionWarning(Identifier);
+
+    if (Identifier.getIdentifierInfo()->getName() == "INFINITY")
```
What about `NAN`?

AaronBallman replied:
Because there are other C standard libraries which do something different. For example, musl: https://git.musl-libc.org/cgit/musl/tree/include/math.h#n18

andykaylor replied:
> Since NAN is defined as (-(float)(INFINITY * 0.0F)) when used in the program it will trigger the "use of infinity via a macro results in undefined behavior due to the currently enabled floating-point options [-Wnan-and-infinity-disabled]". Wouldn't that be enough? If I add NaN here we would get 2 warnings.

This is actually a bit unfortunate. If I use "-fhonor-nans -fno-honor-infinities" this definition of NAN will trigger a warning, even though I expect that we'll constant-fold it as expected without regard to the "no-honor-infinities" setting.

AaronBallman replied:
We will?
```
#ifndef _HUGE_ENUF
    #define _HUGE_ENUF  1e+300  // _HUGE_ENUF*_HUGE_ENUF must overflow
#endif

#define INFINITY   ((float)(_HUGE_ENUF * _HUGE_ENUF))
...
#define NAN        ((float)(INFINITY * 0.0F))
```
the multiplication for `INFINITY` won't form an actual infinity value in that case, so why would the multiplication with `0.0` not generate a NAN macro that folds to 0.0F?

If use of `INFINITY` is undefined behavior when `-fno-honor-infinities` is passed... it makes sense to me that the definition of `NAN` using `INFINITY` is UB by the same logic.

zahiraam committed:
Fixed diag's comment and added NAN to the macro testing.

AaronBallman commented on clang/include/clang/Basic/DiagnosticCommonKinds.td:
```diff
@@ -70,6 +70,10 @@ def warn_pragma_debug_missing_argument : Warning<
 def warn_pragma_debug_unexpected_argument : Warning<
   "unexpected argument to debug command">, InGroup<IgnoredPragmas>;
 
+def warn_fp_nan_inf_when_disabled : Warning<
+  "use of %select{infinity|NaN}0%select{| via a macro}1 results in undefined behavior "
+  "due to the currently enabled floating-point options">,
+  InGroup<DiagGroup<"nan-and-infinity-disabled">>;
```
```suggestion
def warn_fp_nan_inf_when_disabled : Warning<
  "use of %select{infinity|NaN}0%select{| via a macro}1 is undefined behavior "
  "due to the currently enabled floating-point options">,
  InGroup<DiagGroup<"nan-infinity-disabled">>;
```
Sorry for still puttering around with what color to paint the bikeshed, but I think this might be slightly more concise wording and a somewhat better diagnostic group.

AaronBallman commented on clang/test/Sema/warn-infinity-nan-disabled-lnx.cpp:
```diff
@@ -0,0 +1,273 @@
+
+int compareit(float a, float b) {
+  volatile int i, j, k, l, m, n, o, p;
+#if FAST
+// expected-warning@+5 {{use of infinity via a macro results in undefined behavior due to the currently enabled floating-point options}}
+#endif
+#if NO_INFS
+// expected-warning@+2 {{use of infinity via a macro results in undefined behavior due to the currently enabled floating-point options}}
+#endif
```
Rather than use the preprocessor to have conditional diagnostics, it's better to use diagnostics with different prefixes. e.g.,
```
// RUN: %clang_cc1 -x c++ -verify=no-inf,no-nan -triple powerpc64le-unknown-unknown %s \
// RUN: -menable-no-infs -menable-no-nans 

// RUN: %clang_cc1 -x c++ -verify=no-inf -triple powerpc64le-unknown-unknown %s \
// RUN: -menable-no-infs

// RUN: %clang_cc1 -x c++ -verify=no-nan -triple powerpc64le-unknown-unknown %s \
// RUN: -menable-no-nans

...

i = a == INFINITY; // no-inf-warning {{blah blah blah}}
```

zahiraam committed:
Fixed LIT tests, RN and warning message.

AaronBallman commented on clang/test/Sema/warn-infinity-nan-disabled-lnx.cpp:
```diff
@@ -0,0 +1,178 @@
+// RUN: %clang_cc1 -x c++ -verify=no-inf-no-nan -triple powerpc64le-unknown-unknown %s \
+// RUN: -menable-no-infs -menable-no-nans
+
+// RUN: %clang_cc1 -x c++ -verify -triple powerpc64le-unknown-unknown %s \
+// RUN: -DNOFAST=1
+
+// RUN: %clang_cc1 -x c++ -verify=no-inf -triple powerpc64le-unknown-unknown %s \
+// RUN: -menable-no-infs
+
+// RUN: %clang_cc1 -x c++ -verify=no-nan -triple powerpc64le-unknown-unknown %s \
+// RUN: -menable-no-nans
+
+int isunorderedf (float x, float y);
+#if NOFAST
+// expected-no-diagnostics
+#endif
```
```suggestion
// RUN: %clang_cc1 -x c++ -verify=no-fast -triple powerpc64le-unknown-unknown %s

// RUN: %clang_cc1 -x c++ -verify=no-inf -triple powerpc64le-unknown-unknown %s \
// RUN: -menable-no-infs

// RUN: %clang_cc1 -x c++ -verify=no-nan -triple powerpc64le-unknown-unknown %s \
// RUN: -menable-no-nans

// no-fast-no-diagnostics

int isunorderedf (float x, float y);
```

AaronBallman commented on clang/test/Sema/warn-infinity-nan-disabled-win.cpp:
```diff
@@ -0,0 +1,181 @@
+
+int isunorderedf (float x, float y);
+#if NOFAST
+// expected-no-diagnostics
```
Same thing here as with the other test.

AaronBallman reviewed (approved):
The changes LGTM aside from some minor improvements to the tests.

zahiraam committed:
Fix LIT tests.

zahiraam commented:
Any other comments from @arsenm and @andykaylor?

andykaylor reviewed (approved):
I'm happy with this.

mikaelholmen commented:
Hello,

It seems like with this change it warns like
```
w.c:1:10: warning: use of infinity via a macro is undefined behavior due to the currently enabled floating-point options [-Wnan-infinity-disabled]
    1 | # ifndef INFINITY
      |          ^
1 warning generated.
```
with
```
clang w.c -c -ffast-math
```
on input like
```
# ifndef INFINITY

# endif
```
Is this expected and wanted?

AaronBallman commented:
> Is this expected and wanted?

Good catch! I would not expect that diagnostic; we should silence the diagnostic if it's used in a `#ifdef`, `#elifdef`, or `defined` because the value isn't necessary. Perhaps we should also silence something like `sizeof(INFINITY)` because it's unevaluated?

mikaelholmen commented:
> > > Is this expected and wanted?
> > 
> > 
> > Good catch! I would not expect that diagnostic; we should silence the diagnostic if it's used in a `#ifdef`, `#elifdef`, or `defined` because the value isn't necessary. Perhaps we should also silence something like `sizeof(INFINITY)` because it's unevaluated?
> 
> Thanks @mikaelholmen and @AaronBallman. I will take a look and create a PR to fix the issue.

Hi @zahiraam !
Did you get anywhere with this?

zeroomega commented:
Hi, I am seeing breakages on some of our audio code on tip of tree Clang and I traced it back to this patch.

We uses `-ffast-math` mode for performance reasons but we also uses `-fhonor-infinities`, `-fhonor-nans` and `fno-finite-math-only` to ensure that the behavior of infinity is correct. But the message `error: use of infinity is undefined behavior due to the currently enabled floating-point options [-Werror,-Wnan-infinity-disabled] ` was still produced by the compiler. Is this intended behavior? If not, could you revert, revise and reland the change please?

Thanks.

zahiraam commented:
> Hi, I am seeing breakages on some of our audio code on tip of tree Clang and I traced it back to this patch.
> 
> We uses `-ffast-math` mode for performance reasons but we also uses `-fhonor-infinities`, `-fhonor-nans` and `fno-finite-math-only` to ensure that the behavior of infinity is correct. But the message `error: use of infinity is undefined behavior due to the currently enabled floating-point options [-Werror,-Wnan-infinity-disabled] ` was still produced by the compiler. Is this intended behavior? If not, could you revert, revise and reland the change please?
> 
> Thanks.

I think this must be related to the comment above. I am in process of creating a PR to fix it. But if it can't wait I will revert this.
See https://godbolt.org/z/r154xP11W

AaronBallman commented:
> Hi, I am seeing breakages on some of our audio code on tip of tree Clang and I traced it back to this patch.
> 
> We uses `-ffast-math` mode for performance reasons but we also uses `-fhonor-infinities`, `-fhonor-nans` and `fno-finite-math-only` to ensure that the behavior of infinity is correct. But the message `error: use of infinity is undefined behavior due to the currently enabled floating-point options [-Werror,-Wnan-infinity-disabled] ` was still produced by the compiler. Is this intended behavior? If not, could you revert, revise and reland the change please?
> 
> Thanks.

What is the order you're passing the flags? This may have actually pointed out a difference between what you expected and what actually happens (which is part of the point to these changes): https://godbolt.org/z/x54qTGoMG

AaronBallman commented:
> > > > > Is this expected and wanted?
> > > > 
> > > > 
> > > > Good catch! I would not expect that diagnostic; we should silence the diagnostic if it's used in a `#ifdef`, `#elifdef`, or `defined` because the value isn't necessary. Perhaps we should also silence something like `sizeof(INFINITY)` because it's unevaluated?
> > > 
> > > 
> > > Thanks @mikaelholmen and @AaronBallman. I will take a look and create a PR to fix the issue.
> > 
> > 
> > Hi @zahiraam ! Did you get anywhere with this?
> 
> Sorry! I didn't have time to work on it yet. I should probably get to it next week.

Because these changes made it into 18.x and we're already seeing feedback about the diagnostic quality, I think we need to either revert the changes for the moment (both on main and on 18.x) or we need to commit to getting the fix as quickly as possible (and backport it to 18.x). I'd like to avoid getting into a situation where we're reverting this on 18.x later in the cycle. How about this for a plan: if there's not a fix committed by EOD Mon (2/5), we revert so that rc2 goes out in a good state (rc2 is being cut on Tue the 6th)?

zeroomega commented:
> > Hi, I am seeing breakages on some of our audio code on tip of tree Clang and I traced it back to this patch.
> > We uses `-ffast-math` mode for performance reasons but we also uses `-fhonor-infinities`, `-fhonor-nans` and `fno-finite-math-only` to ensure that the behavior of infinity is correct. But the message `error: use of infinity is undefined behavior due to the currently enabled floating-point options [-Werror,-Wnan-infinity-disabled] ` was still produced by the compiler. Is this intended behavior? If not, could you revert, revise and reland the change please?
> > Thanks.
> 
> What is the order you're passing the flags? This may have actually pointed out a difference between what you expected and what actually happens (which is part of the point to these changes): https://godbolt.org/z/x54qTGoMG

Thanks for the tip.
I think it is the ordering issue on our side. The build system actually put 2 `-ffast-math` flags into the command line (we expect 1 but some the buildfile have a bug), The build system placed `-fhonor-infinities, -fhonor-nans` before the 2nd `-ffast-math` flag and this 2nd `--ffast-math` resets those honor flags I assume, and we saw the warning.
